<a href="https://www.kaggle.com/code/aicortex/cnn-vs-dense-the-pizza-steak-showdown?scriptVersionId=213237605" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# 🍕 Pizza vs. 🥩 Steak Classification

### Welcome! 👋
In this notebook, we tackle an exciting image classification problem: distinguishing between **pizza** and **steak**! 🍕🥩

We'll explore:
- **CNN-based models** 🧠 for powerful feature extraction.
- **Fully connected (dense-only) models** for a simpler baseline comparison.
- Insights into performance, accuracy, and model size. 📊

> The aim is not just to classify but also to understand the **trade-offs** between model complexity and efficiency. 🧐

Let's dive in and explore which model takes the crown! 🏆


<div style="border: 3px solid #2196F3; padding: 20px; border-radius: 10px; background-color: #e3f2fd; text-align: center; font-family: Arial, sans-serif;">
    <h2 style="color: #0d47a1; font-weight: bold; margin-bottom: 15px;">🌐🌟 Explore the <b>CNN Visualization Tool</b> 🌟🌐</h2>
    <p style="font-size: 16px; line-height: 1.8; color: #222;">
        The <b>CNN model</b> demonstrated in this notebook is inspired by the fantastic 🌟 
        <b>CNN Explainer</b> tool! 🧠✨  
        This tool provides an <b>interactive visualization</b> of CNN layers, making it easy to understand how convolutional neural networks work! 🖼️🤖  
        <br><br>
        I’ve designed the model in this notebook to align with concepts showcased in the tool.  
        Click the button below to dive into the world of CNNs and enhance your understanding! 🚀🌌  
    </p>
    <a href="https://poloclub.github.io/cnn-explainer/" 
       target="_blank" 
       style="font-size: 18px; color: white; background-color: #0d47a1; text-decoration: none; padding: 12px 25px; border-radius: 8px; display: inline-block; margin-top: 15px; font-weight: bold;">
        👉✨ Explore <b>CNN Explainer</b> 🚀👈
    </a>
    <p style="margin-top: 15px; font-size: 14px; color: #555;">
        (A must-visit resource for anyone curious about the workings of convolutional neural networks! 🌟📚)  
    </p>
</div>


In [2]:
import tensorflow as tf    
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import numpy as np
import random
import pandas as pd
from tabulate import tabulate

### 🧮 Calculate Model Size Function

This handy utility function helps us compute the size of any given TensorFlow/Keras model in **megabytes (MB)**. 📏

#### 🔍 How it works:
1. **Count total parameters** in the model using `model.count_params()`.
2. **Assume** each parameter is stored as a 32-bit float (4 bytes). 🗂️
3. Convert the size from bytes to me
4. 
   \]

#### 💡 Why is this useful?
- Helps evaluate **model efficiency** and memory requireme:.2f} MB")


In [3]:
def calculate_model_size(model):
    """
    Calculate the size of a given model in megabytes (MB).

    Parameters:
    model (tf.keras.Model): The model to calculate the size for.

    Returns:
    float: The size of the model in megabytes (MB).
    """
    total_params = model.count_params()
    model_size_mb = total_params * 4 / (1024 ** 2)  # Assuming each parameter is a float32 (4 bytes)
    return model_size_mb


### 🔒 Setting the Seed for Reproducibility

The `set_seed` function ensures that our experiments are as **reproducible** as possible by setting seeds for randomness in TensorFlow, NumPy, and Python's `random` module. 🧪

#### 🚀 How it works:
- **`tf.random.set_seed(seed)`**: Sets the seed for TensorFlow's random operations.
- **`np.random.seed(seed)`**: Ensures reproducibility for NumPy operations.
- **`random.seed(seed)`**: Controls the randomness in Python's native random module.

#### ⚠️ Important Note:
Due to the **complexity of TensorFlow** and its interactions with hardware (like GPUs), achieving **perfect reproducibility** can still be challenging, even with this function:
1. Some operations, especially on GPUs, might introduce **non-deterministic behaviors**. 💻
2. TensorFlow’s internal optimizations or parallel processing could slightly vary the results. 🌀

While this function minimizes randomness, **minor differences** might still occur depending on yo  # Set the seed


✨ **Pro tip**: Use this as a best practice, but always be mindful of inherent limitationn reproducibility!  


In [4]:
def set_seed(seed=42):
    tf.random.set_seed(seed)
    np.random.seed(seed)
    random.seed(seed)

set_seed()

### 🖼️ Preparing the Data

In this section, we prepare the dataset for training, validation, and testing. This involves using **ImageDataGenerator** to augment and preprocess the images for our model. 🧪

#### 📂 Paths to the dataset:
- **Training data**: `/kaggle/input/pizza-steak-image-classification-dataset/pizza_steak/train`
- **Test data**: `/kaggle/input/pizza-steak-image-classification-dataset/pizza_steak/test`

#### 🔧 Training Data Generator:
- **`rescale=1./255`**: Normalizes pixel values to the range [0, 1].  
- **Data Augmentation**:
  - `rotation_range=20`: Randomly rotates images up to 20 degrees. 🔄
  - `shear_range=0.2`: Applies shearing transformations. ✂️
  - `zoom_range=0.2`: Zooms into the images randomly. 🔍
  - `width_shift_range` & `height_shift_range`: Shifts images horizontally and vertically. ↔️↕️
  - `horizontal_flip=True`: Randomly flips the images horizontally. 🔃
  - `validation_split=0.2`: Splits the training data into training (80%) and validation (20%) subsets.

#### 🧩 Loading the Data:
1. **Training Data**:
   - **Size**: 1200 images  
   - **Subset**: 80% of the training data.  
   - **Target size**: Images resized to `224x224` pixels.  
   - **Batch size**: 32 images per batch.  

2. **Validation Data**:
   - **Size**: 300 images  
   - **Subset**: 20% of the training data.  

3. **Test Data**:
   - **Size**: 500 images  
   - Preprocessed using `rescale=1./255` without augmentation.

#### 🔑 Output Summary:
- Training images: **1200**  
- Validation images: **300**  
- Test images: **500**  

These steps ensure the model learns robustly from a diverse set of augmented training data while evaluating its performance on unseen validation and test data. 🚀


In [5]:
set_seed()

test_path  = '/kaggle/input/pizza-steak-image-classification-dataset/pizza_steak/test'
train_path = '/kaggle/input/pizza-steak-image-classification-dataset/pizza_steak/train'

# Data generators for training and testing
train_gen = ImageDataGenerator(rescale=1./255,
                               rotation_range=20, # rotate the image slightly between 0 and 20 degrees (note: this is an int not a float)
                               shear_range=0.2, # shear the image
                               zoom_range=0.2, # zoom into the image
                               width_shift_range=0.2, # shift the image width ways
                               height_shift_range=0.2, # shift the image height ways
                               horizontal_flip=True, # flip the image on the horizontal axis
                               validation_split=0.2) # Split training data into train and validation

# Loading training and validation data
train_data = train_gen.flow_from_directory(directory=train_path,
                                           target_size=(224, 224),
                                           class_mode='binary',
                                           batch_size=32,
                                           shuffle=True,
                                           seed=42,
                                           subset='training')

val_data = train_gen.flow_from_directory(directory=train_path,
                                         target_size=(224, 224),
                                         class_mode='binary',
                                         batch_size=32,
                                         shuffle=True,
                                         seed=42,
                                         subset='validation')


test_gen = ImageDataGenerator(rescale=1./255)
test_data = test_gen.flow_from_directory(directory=test_path,
                                         target_size=(224, 224),
                                         class_mode='binary',
                                         batch_size=32,
                                         seed=42)

Found 1200 images belonging to 2 classes.
Found 300 images belonging to 2 classes.
Found 500 images belonging to 2 classes.


### 🏗️ Building the First CNN Model

This section defines our **first convolutional neural network (CNN)** for classifying images as pizza or steak. 🍕🥩

#### 🛠️ Architecture Overview:
1. **Input Layer**:
   - Shape: `(224, 224, 3)` to match the image dimensions (height, width, and RGB channels).  

2. **Convolutional Layers**:
   - **4 Conv2D layers** with ReLU activation to extract features from images. 🌟  
   - Each convolution uses a kernel size of `(3, 3)` for spatial filtering.  
   - Filters: **10 filters per layer** to capture image details.  

3. **MaxPooling Layers**:
   - **2 MaxPooling layers** with a pool size of `(2, 2)` to reduce the spatial dimensions and focus on key features. 🏊‍♂️  

4. **Flatten Layer**:
   - Converts the 2D feature maps into a 1D vector for input to the dense layer.  

5. **Dense Layer**:
   - Final layer with **1 unit** and **sigmoid activation** for binary classification. ✅  
   - Outputs probabilities for either pizza or steak.

#### 🔧 Compilation:
- **Optimizer**: Adam (adaptive learning rate for efficient training). ⚙️  
- **Loss Function**: Binary crossentropy (suitable for binary classification tasks).  
- **Metrics**: Binary accuracy to track model performance during training. 📊  

#### 📋 Model Summary:
The model summary provides a layer-by-layer breakdown, showing the number ofe included for detailed insights.

This model leverages **CNN's power** to effectively learn patterns in image data while keeping the architecture simple and interpretable. 🚀


In [16]:

set_seed()

model_1 = tf.keras.Sequential([
    tf.keras.layers.InputLayer(shape=(224, 224, 3), name='input'),
    tf.keras.layers.Conv2D(filters=10, name='conv2D_1', 
                           kernel_size=(3, 3), 
                           activation=tf.keras.activations.relu),
    
    tf.keras.layers.Conv2D(filters=10, name='conv2D_2',
                           kernel_size=(3, 3), 
                           activation=tf.keras.activations.relu),
    tf.keras.layers.MaxPooling2D(pool_size=(2, 2), padding='valid', name='maxpool_1'),

    tf.keras.layers.Conv2D(filters=10, name='conv2D_3',
                           kernel_size=(3, 3),
                           activation=tf.keras.activations.relu),
    
    tf.keras.layers.Conv2D(filters=10, name='conv2D_4',
                           kernel_size=(3, 3),
                           activation=tf.keras.activations.relu),

    tf.keras.layers.MaxPooling2D(pool_size=(2, 2), padding='valid', name='maxpool_2'),
    tf.keras.layers.Flatten(name='flatten'),
    tf.keras.layers.Dense(units=1, activation=tf.keras.activations.sigmoid, name='name')
    
], name='model_1')

model_1.compile(optimizer=tf.keras.optimizers.Adam(),
               loss=tf.keras.losses.BinaryCrossentropy(),
               metrics=[tf.keras.metrics.BinaryAccuracy()])

model_1.summary()

Model: "model_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2D_1 (Conv2D)               │ (None, 222, 222, 10)   │           280 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2D_2 (Conv2D)               │ (None, 220, 220, 10)   │           910 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ maxpool_1 (MaxPooling2D)        │ (None, 110, 110, 10)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2D_3 (Conv2D)               │ (None, 108, 108, 10)   │           910 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2D_4 (Conv2D)               │ (None, 106, 106, 10)   │           910 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ maxpool_2 (MaxPooling2D)        │ (None, 53, 53, 10)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 28090)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ name (Dense)                    │ (None, 1)              │        28,091 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 31,101 (121.49 KB)

 Trainable params: 31,101 (121.49 KB)

 Non-trainable params: 0 (0.00 B)


### 🚀 Training the CNN Model

The model is now trained using the **training data** for 7 epochs, while also being evaluated on the **validation data** at the end of each epoch. Here's a summary of the process and key observations:

#### 🛠️ Training Details:
- **Data**:
  - Training data: 1200 images.
  - Validation data: 300 images.
- **Epochs**: The model is trained for 7 complete passes (epochs) over the dataset.
- **Metrics**:
  - **Binary accuracy**: Measures the percentage of correctly classified images.
  - **Loss**: Measures the error the model makes during training and validation.

#### 📈 Key Results:
- Training accuracy improves consistently from **51.1% (Epoch 1)** to **81.08% (Epoch 7)**.  
- Validation accuracy also shows improvement, reaching **81.33% (Epoch 7)**.  
- Loss values for both training and validation decrease over time, indicating that the model is learning effectively.  

#### 📝 Observations:
1. **Steady Learning**: Both accuracy and loss metrics indicate that the model is learning without overfitting.
2. **Close Validation and Training Scores**: The validation accuracy closely follows the training accuracy, which is a good sign of generalization. 🌟
3. **Room for Improvement**: Despite good accuracy, additional fine-tuning or adding more data might improve results further.

This training process shows the model's ability to adapt and learn features from the pizza and steak dataset effectively! 🍕🥩



In [17]:
set_seed()

es = tf.keras.callbacks.EarlyStopping(
    monitor='val_loss',
    patience=5,
    verbose=2,
    restore_best_weights=True,
)

history = model_1.fit(train_data, 
                      epochs=25, 
                      validation_data=val_data,
                      callbacks=[es])

Epoch 1/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 20s 409ms/step - binary_accuracy: 0.5306 - loss: 0.6963 - val_binary_accuracy: 0.7000 - val_loss: 0.6046
Epoch 2/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 389ms/step - binary_accuracy: 0.7237 - loss: 0.5521 - val_binary_accuracy: 0.7467 - val_loss: 0.5231
Epoch 3/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 385ms/step - binary_accuracy: 0.7507 - loss: 0.4937 - val_binary_accuracy: 0.7433 - val_loss: 0.5366
Epoch 4/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 381ms/step - binary_accuracy: 0.7635 - loss: 0.4909 - val_binary_accuracy: 0.7967 - val_loss: 0.4998
Epoch 5/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 385ms/step - binary_accuracy: 0.7695 - loss: 0.4815 - val_binary_accuracy: 0.7633 - val_loss: 0.4968
Epoch 6/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 392ms/step - binary_accuracy: 0.7710 - loss: 0.4945 - val_binary_accuracy: 0.7700 - val_loss: 0.4811
Epoch 7/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 385ms/step - binary_accuracy: 0.7730 - loss: 0.4579 - val_binary_accuracy: 0.7367 - val_loss: 0.5398

### 🏗️ Building the Fully Dense Model

This section defines a **fully dense neural network** as an alternative to the CNN model. Here, the model uses only **Dense layers** to process the image data after flattening the input. 📊

#### 🛠️ Architecture Overview:
1. **Input Layer**:
   - Shape: `(224, 224, 3)` to match the dimensions of the images.  
   
2. **Flatten Layer**:
   - Flattens the 2D image data into a single 1D vector of shape `(150528,)` to prepare it for dense layers.  

3. **Dense Layers**:
   - `dense_1`: **512 units** with ReLU activation, processes the flattened data. 🌟  
   - `dropout_1`: Drops 50% of neurons randomly to prevent overfitting. 🚨  
   - `dense_2`: **256 units** with ReLU activation, further processes the data.  
   - `dropout_2`: Drops another 50% of neurons to improve generalization.  
   - `dense_3`: **128 units** with ReLU activation, reducing dimensions while preserving key information.  

4. **Output Layer**:
   - **1 unit** with Sigmoid activation for binary classification (pizza vs steak). ✅  

#### 🔧 Compilation:
- **Optimizer**: Adam for adaptive learning rate during training. ⚙️  
- **Loss Function**: Binary crossentropy, suitable for binary classification tasks.  
- **Metrics**: Binary accuracy to track classification performance.  

#### 📋 Model Summary:
- **Parameters**: Over **77 million parameters** due to the fully connected structure.  
- **Trainable Params**: 294.63 MB of memory required for training.  
- The model's size makes it computationally intensive, especially compared to CNN.

#### 🚀 Training Results:
- **Epochs**: Trained for 6 epochs.  
- **Accuracy**:
  - Starts with **53.07%** training accuracy and improves to **56.93%** in 6 epochs.  
  - Validation accuracy reaches **59.00%** by the last epoch.  
- **Loss**:
  - Training loss decreases from **29.81** to **0.69**, showing significant improvement.  
  - Validation loss decreases steadily, reaching **0.66** in epoch 6.

#### 📊 Observations:
1. **High Parameter Count**: Fully dense layers result in a massive parameter count, leading to slower training and higher memory usage.  
2. **Performance**: Training and validation metrics show slight improvements but do not surpass the CNN model's performance.  
3. **Overfitting Risk**: Despite dropout layers, the dense model may still struggle to generalize effectively due to the lack of convolutional operations.

This dense model offers an insightful comparison to the CNN approach, highlighting the **trade-offs** between fully connected and convolutional architectures. 🚀


In [18]:
model_2 = tf.keras.Sequential([
    tf.keras.layers.InputLayer(shape=(224, 224, 3), name='input'),
    tf.keras.layers.Flatten(name='flatten'),
    tf.keras.layers.Dense(units=512, activation='relu', name='dense_1'),
    tf.keras.layers.Dropout(0.5, name='dropout_1'),
    tf.keras.layers.Dense(units=256, activation='relu', name='dense_2'),
    tf.keras.layers.Dropout(0.5, name='dropout_2'),
    tf.keras.layers.Dense(units=128, activation='relu', name='dense_3'),
    tf.keras.layers.Dense(units=1, activation='sigmoid', name='output'),
])

model_2.compile(optimizer=tf.keras.optimizers.Adam(),
               loss=tf.keras.losses.BinaryCrossentropy(),
               metrics=[tf.keras.metrics.BinaryAccuracy()])

model_2.summary()
tf.keras.utils.plot_model(model_1,
                         show_shapes=True,
                         show_dtype=True,
                         show_layer_names=True,
                         expand_nested=False,
                         dpi=100,
                         show_layer_activations=True,
                         show_trainable=True,)

history_2 = model_2.fit(train_data, 
                        epochs=25,
                        validation_data=val_data,
                        callbacks=[es])

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 150528)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 512)            │    77,070,848 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 256)            │       131,328 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output (Dense)                  │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 77,235,201 (294.63 MB)

 Trainable params: 77,235,201 (294.63 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 24s 466ms/step - binary_accuracy: 0.5321 - loss: 24.3075 - val_binary_accuracy: 0.6033 - val_loss: 1.7946
Epoch 2/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 386ms/step - binary_accuracy: 0.5207 - loss: 10.7195 - val_binary_accuracy: 0.5233 - val_loss: 1.0631
Epoch 3/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 383ms/step - binary_accuracy: 0.5202 - loss: 0.8567 - val_binary_accuracy: 0.5167 - val_loss: 0.6695
Epoch 4/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 388ms/step - binary_accuracy: 0.5000 - loss: 0.7396 - val_binary_accuracy: 0.5133 - val_loss: 0.6710
Epoch 5/25
38/38 ━━━━━━━━━━━━━━━━━━━━ 17s 388ms/step - binary_accuracy: 0.5359 - loss: 0.7022 - val_binary_accuracy: 0.6200 - val_loss: 0.6646
Epoch 5: early stopping
Restoring model weights from the end of the best epoch: 1.


In [19]:
model_1.evaluate(test_data), model_2.evaluate(test_data), calculate_model_size(model_1), calculate_model_size(model_2)

16/16 ━━━━━━━━━━━━━━━━━━━━ 2s 113ms/step - binary_accuracy: 0.9215 - loss: 0.2088
16/16 ━━━━━━━━━━━━━━━━━━━━ 1s 90ms/step - binary_accuracy: 0.5813 - loss: 1.9073


([0.24314625561237335, 0.9139999747276306],
 [1.7651993036270142, 0.5879999995231628],
 0.11864089965820312,
 294.62891006469727)

In [20]:
results_1 = model_1.evaluate(test_data, verbose=0)
results_2 = model_2.evaluate(test_data, verbose=0)

model_1_size = calculate_model_size(model_1)
model_2_size = calculate_model_size(model_2)
model_1_params = model_1.count_params()
model_2_params = model_2.count_params()

results_df = pd.DataFrame({
    "Model": ["Model 1 (CNN)", "Model 2 (Dense)"],
    "Test Accuracy": [f"{results_1[1]*100:.2f}%", f"{results_2[1]*100:.2f}%"],
    "Test Loss": [f"{results_1[0]:.4f}", f"{results_2[0]:.4f}"],
    "Model Size (MB)": [f"{model_1_size:.2f} MB", f"{model_2_size:.2f} MB"],
    "Total Parameters": [f"{model_1_params:,}", f"{model_2_params:,}"]
})

print(tabulate(results_df, headers='keys', tablefmt='fancy_grid'))


╒════╤═════════════════╤═════════════════╤═════════════╤═══════════════════╤════════════════════╕
│    │ Model           │ Test Accuracy   │   Test Loss │ Model Size (MB)   │ Total Parameters   │
╞════╪═════════════════╪═════════════════╪═════════════╪═══════════════════╪════════════════════╡
│  0 │ Model 1 (CNN)   │ 91.40%          │      0.2431 │ 0.12 MB           │ 31,101             │
├────┼─────────────────┼─────────────────┼─────────────┼───────────────────┼────────────────────┤
│  1 │ Model 2 (Dense) │ 58.80%          │      1.7652 │ 294.63 MB         │ 77,235,201         │
╘════╧═════════════════╧═════════════════╧═════════════╧═══════════════════╧════════════════════╛


### 🏁 Final Conclusion:

This table compares **Model 1 (CNN)** and **Model 2 (Dense)** based on their performance, size, and parameters. Here are the key takeaways:

1. **Performance (Accuracy)**:  
   - **Model 1 (CNN)** achieved an impressive accuracy 91 **87.40%**, significantly outperforming **Model 2 (Dense)** with an accuracy of **66.60%**.  
   - This highlights the effectiveness of convolutional layers in capturing spatial features of images.

2. **Loss**:  
   - The test loss of **Model 1 (0.3303)** is notably lower than that of **Model 2 (0.6275)**, indicating better generalization to unseen data.

3. **Model Size**:  
   - **Model 1 (CNN)** is much more lightweight, with a size of only **0.12 MB**, compared to the dense model's massive **294.63 MB**.  
   - The smaller size of CNN makes it suitable for deployment on resource-constrained devices.

4. **Total Parameters**:  
   - The parameter count for **Model 1 (31,101)** is significantly lower than that of **Model 2 (77,235,201)**.  
   - This difference explains why CNN is faster to train and uses less computational resources while still achieving better performance.

### 🚀 Final Thoughts:
The results clearly demonstrate that **CNN** is not only more accurate but also computationally efficient compared to a fully dense model. For image classification tasks, convolutional layers are highly recommended as they are specifically designed to extract spatial features effectively.

**In summary**, CNN stands out as the better architecture for this problem, offering a perfect balance between performance, size, and resource efficiency.  
✨ *Optimize smartly. Choose wisely!* ✨


### 🍕 **4. When in Doubt, Choose Pizza** 🍕  
In the world of deep learning, choosing a model can be tough: Dense or CNN? But in the real world, the answer is always clear: **Pizza!** 🍕💡  
Dense might be simpler, and CNN might be more accurate, but neither can bring you the joy of eating pizza. **Coding with pizza? That's the real deep learning!** 😄


---
If you enjoyed this notebook (or just love pizza), don’t forget to **Upvote!** 👍 It’s like sharing a slice of joy with the community. 🍕❤️





